File: src/main/kotlin/com/hanghea99/commerce/configuration/database/ReplicationRoutingDataSource.java
package com.hanghea99.commerce.configuration.database;

import java.util.HashMap;
import java.util.Map;
import javax.sql.DataSource;
import org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource;
import org.springframework.transaction.support.TransactionSynchronizationManager;

public class ReplicationRoutingDataSource extends AbstractRoutingDataSource {
    @Override
    protected Object determineCurrentLookupKey() {
        return TransactionSynchronizationManager.isCurrentTransactionReadOnly()
                ? DataSourceType.SLAVE
                : DataSourceType.MASTER;
    }

    public ReplicationRoutingDataSource(DataSource masterDataSource, Object slaveDataSource) {
        Map<Object, Object> dataSourceMap = new HashMap<>();
        dataSourceMap.put(DataSourceType.MASTER, masterDataSource);
        dataSourceMap.put(DataSourceType.SLAVE, slaveDataSource);

        setTargetDataSources(dataSourceMap);
        setDefaultTargetDataSource(masterDataSource);
    }
}

File: src/main/kotlin/com/hanghea99/commerce/configuration/database/ReplicationRoutingDataSource.kt
package com.hanghea99.commerce.configuration.database

import org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource
import org.springframework.transaction.support.TransactionSynchronizationManager
import javax.sql.DataSource

class ReplicationRoutingDataSource(masterDataSource: DataSource, slaveDataSource: Any) :
    AbstractRoutingDataSource() {
    override fun determineCurrentLookupKey(): Any? {
        return if (TransactionSynchronizationManager.isCurrentTransactionReadOnly()) DataSourceType.SLAVE else DataSourceType.MASTER
    }

    init {
        val dataSourceMap: MutableMap<Any, Any> = HashMap()
        dataSourceMap[DataSourceType.MASTER] = masterDataSource
        dataSourceMap[DataSourceType.SLAVE] = slaveDataSource
        setTargetDataSources(dataSourceMap)
        setDefaultTargetDataSource(masterDataSource)
    }
}